### PR TITLE
Add context to precompile call.

### PR DESCRIPTION
**Pull request by shaopengw**

### CODE REVIEW

File: runtime/mandala/src/lib.rs [hunk 1]
```diff
@@ -1244,13 +1246,28 @@ impl pallet_contracts::Trait for Runtime {
...
+			s.insert(H160::from_low_u64_be(PREDEPLOY_ADDRESS_START + i));
+		}
+		s
+	};
+}
+
+pub struct MultiCurrencyPrecompileCallerFilter;
+impl PrecompileCallerFilter for MultiCurrencyPrecompileCallerFilter {
+	fn is_allowed(caller: H160) -> bool {
+		MultiCurrencyPrecompileAllowedCallers::get().contains(&caller)
```
[the remark this answers is not in the document]
Comment (shaopengw) [reply runtime/mandala/src/lib.rs]: Yes, if that's safe enough for us.

File: runtime/common/src/precompile/mod.rs [hunk 2]
```diff
@@ -33,12 +33,13 @@ where
 		address: H160,
 		input: &[u8],
 		target_gas: Option<usize>,
+		context: &Context,
 	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>> {
-		EthereumPrecompiles::execute(address, input, target_gas).or_else(|| {
+		EthereumPrecompiles::execute(address, input, target_gas, context).or_else(|| {
 			if address == H160::from_low_u64_be(PRECOMPILE_ADDRESS_START) {
```
Comment (xlc) [inline runtime/common/src/precompile/mod.rs:39]: maybe move the filter check to here?

**Review by xlc — APPROVED** [review]

File: runtime/common/src/precompile/mod.rs (hunk 2 above)
Comment (shaopengw) [reply runtime/common/src/precompile/mod.rs:39]: We might need custom configurable allowed list for other precompiles, to support non-predeployed contracts. Like the NFT contract for linkdrop.

File: runtime/common/src/lib.rs [hunk 3]
```diff
@@ -257,9 +257,32 @@ parameter_types! {
 	};
 }
 
-pub struct MultiCurrencyPrecompileCallerFilter;
-impl PrecompileCallerFilter for MultiCurrencyPrecompileCallerFilter {
+parameter_types! {
+	pub const ZeroBytesLength: usize = 12;
```
Comment (xlc) [inline runtime/common/src/lib.rs:261]: a normal const should be enough?

Comment (shaopengw) [reply runtime/common/src/lib.rs:261]: emmm, yep, will update

**Pull request merged**